PR: https://github.com/odoo/odoo/pull/234974

From: 113d77eb35aa8f8b503043d1201b6cdf78e9be83
From: Hesham Saleh
Date: 2025-12-24 09:57:26

Breaking data model changes scores: del:10.4 + add:0, change matches:
-    bank_acc_number = fields.Char(related='bank_account_id.acc_number', readonly=False)
-    bank_id = fields.Many2one('res.bank', related='bank_account_id.bank_id', readonly=False)
-    duplicate_bank_partner_ids = fields.Many2many(related="bank_ids.duplicate_bank_partner_ids")
+    duplicate_bank_partner_ids = fields.Many2many('res.partner', compute='_compute_duplicate_bank_partner_ids')
-    has_iban_warning = fields.Boolean(
-    partner_country_name = fields.Char(related='partner_id.country_id.name')
-    has_money_transfer_warning = fields.Boolean(
-    money_transfer_service = fields.Char(compute='_compute_money_transfer_service_name')
-    bank_id = fields.Many2one(tracking=True)
-    acc_number = fields.Char(tracking=True)
-    acc_holder_name = fields.Char(tracking=True)
-    currency_id = fields.Many2one(tracking=True)

Total Changes: 210

[IMP] *: Improve UX of bank accounts

This commit improves the UX of bank accounts by:
1. removing res.bank model
2. removes localizations redundancy
3. some fields in localizations should be available to all companies
4. most countries had missing fields

task-5145618

closes odoo/odoo#234974

Related: odoo/enterprise#99418
Related: odoo/upgrade#8958
Related: odoo/upgrade-util#358
Signed-off-by: Habib Ayob (ayh) <ayh@odoo.com>

================================= pseudo patch: =================================

--- a/addons/account/models/account_bank_statement_line.py
+++ b/addons/account/models/account_bank_statement_line.py
@@ -489,12 +489,12 @@ class AccountBankStatementLine(models.Model):
         # However, at the end, we need to filter out the results to not trigger the check_company when trying to
         # assign a res.partner.bank owned by another company.
         bank_account = self.env['res.partner.bank'].sudo().with_context(active_test=False).search([
-            ('acc_number', '=', self.account_number),
+            ('account_number', '=', self.account_number),
             ('partner_id', '=', self.partner_id.id),
         ])
         if not bank_account and not self.env['ir.config_parameter'].sudo().get_bool("account.skip_create_bank_account_on_reconcile"):
             bank_account = self.env['res.partner.bank'].create({
-                'acc_number': self.account_number,
+                'account_number': self.account_number,
                 'partner_id': self.partner_id.id,
                 'journal_id': None,
             })

--- a/addons/account/models/account_journal.py
+++ b/addons/account/models/account_journal.py
@@ -3,7 +3,7 @@ from urllib.parse import urlencode
 
 from odoo import api, Command, fields, models, _
 from odoo.exceptions import UserError, ValidationError
-from odoo.addons.base.models.res_bank import sanitize_account_number
+from odoo.addons.base.models.res_partner_bank import sanitize_account_number
 from odoo.tools import email_normalize, email_normalize_all, groupby, urls
 from odoo.tools.misc import hash_sign
 from collections import defaultdict
@@ -248,8 +248,9 @@ class AccountJournal(models.Model):
         check_company=True,
         domain="[('partner_id','=', company_partner_id)]")
     bank_statements_source = fields.Selection(selection='_get_bank_statements_available_sources', string='Bank Feeds', default='undefined', help="Defines how the bank statements will be registered")
-    bank_acc_number = fields.Char(related='bank_account_id.acc_number', readonly=False)
-    bank_id = fields.Many2one('res.bank', related='bank_account_id.bank_id', readonly=False)
+    bank_name = fields.Char(related='bank_account_id.bank_name', readonly=False)
+    bank_account_number = fields.Char(related='bank_account_id.account_number', readonly=False)
+    bank_bic = fields.Char(related='bank_account_id.bank_bic', readonly=False)
 
     # alias configuration for journals
     alias_name = fields.Char(help="Send one separate email for each invoice.\n"
@@ -737,7 +738,7 @@ class AccountJournal(models.Model):
                 bank_accounts += bank_account
         self.env['account.payment.method.line'].search([('journal_id', 'in', self.ids)]).unlink()
         ret = super(AccountJournal, self).unlink()
-        bank_accounts.unlink()
+        bank_accounts.action_archive()
         return ret
 
     def copy_data(self, default=None):
@@ -790,9 +791,6 @@ class AccountJournal(models.Model):
                         'company_id': company.id,
                         'partner_id': company.partner_id.id,
                     })
-            if 'currency_id' in vals:
-                if journal.bank_account_id:
-                    journal.bank_account_id.currency_id = vals['currency_id']
             if 'bank_account_id' in vals:
                 if vals.get('bank_account_id'):
                     bank_account = self.env['res.partner.bank'].browse(vals['bank_account_id'])
@@ -824,9 +822,9 @@ class AccountJournal(models.Model):
                 journal.default_account_id.currency_id = journal.currency_id
 
         # Create the bank_account_id if necessary
-        if 'bank_acc_number' in vals:
+        if 'bank_account_number' in vals:
             for journal in self.filtered(lambda r: r.type == 'bank' and not r.bank_account_id):
-                journal.set_bank_account(vals.get('bank_acc_number'), vals.get('bank_id'))
+                journal.set_bank_account(vals.get('bank_account_number'), vals.get('bank_bic'))
 
         return result
 
@@ -984,7 +982,7 @@ class AccountJournal(models.Model):
             has_loss_account = vals.get('loss_account_id')
 
             # === Fill missing name ===
-            vals['name'] = vals.get('name') or vals.get('bank_acc_number') or vals.get('name_placeholder')
+            vals['name'] = vals.get('name') or vals.get('bank_account_number') or vals.get('name_placeholder')
 
             # === Fill missing accounts ===
             if not has_liquidity_accounts:
@@ -1034,25 +1032,24 @@ class AccountJournal(models.Model):
 
         for journal, vals in zip(journals, vals_list):
             # Create the bank_account_id if necessary
-            if journal.type == 'bank' and not journal.bank_account_id and vals.get('bank_acc_number'):
-                journal.set_bank_account(vals.get('bank_acc_number'), vals.get('bank_id'))
+            if journal.type == 'bank' and not journal.bank_account_id and vals.get('bank_account_number'):
+                journal.set_bank_account(vals.get('bank_account_number'), vals.get('bank_bic'))
 
         return journals
 
-    def set_bank_account(self, acc_number, bank_id=None):
+    def set_bank_account(self, account_number, bank_bic=None):
         """ Create a res.partner.bank (if not exists) and set it as value of the field bank_account_id """
         self.ensure_one()
         res_partner_bank = self.env['res.partner.bank'].search([
-            ('sanitized_acc_number', '=', sanitize_account_number(acc_number)),
+            ('sanitized_account_number', '=', sanitize_account_number(account_number)),
             ('partner_id', '=', self.company_id.partner_id.id),
         ], limit=1)
         if res_partner_bank:
             self.bank_account_id = res_partner_bank.id
         else:
             self.bank_account_id = self.env['res.partner.bank'].create({
-                'acc_number': acc_number,
-                'bank_id': bank_id,
-                'currency_id': self.currency_id.id,
+                'account_number': account_number,
+                'bank_bic': bank_bic,
                 'partner_id': self.company_id.partner_id.id,
                 'journal_id': self,
             }).id

--- a/addons/account/models/account_move.py
+++ b/addons/account/models/account_move.py
@@ -1041,18 +1041,6 @@ class AccountMove(models.Model):
 
     @api.depends('bank_partner_id', 'currency_id', 'preferred_payment_method_line_id')
     def _compute_partner_bank_id(self):
-        def _bank_selection_key(bank):
-            """Sorting priority:
-            0. Same currency as the move or no currency
-            1. Different currency
-            Then: prefer banks allowing outgoing payments (trusted ones)
-            """
-            if bank.currency_id == move.currency_id or not bank.currency_id:
-                currency_priority = 0
-            else:
-                currency_priority = 1
-            return (currency_priority, not bank.allow_out_payment)
-
         for move in self:
             if move.is_inbound() and (
                 payment_method := (
@@ -1065,7 +1053,7 @@ class AccountMove(models.Model):
 
             move.partner_bank_id = move.bank_partner_id.bank_ids.filtered(
                 lambda bank: not bank.company_id or bank.company_id == move.company_id
-            ).sorted(key=_bank_selection_key)[:1]
+            ).sorted(key=lambda b: not b.allow_out_payment)[:1]
 
     @api.depends('partner_id')
     def _compute_invoice_payment_term_id(self):

--- a/addons/account/models/partner.py
+++ b/addons/account/models/partner.py
@@ -342,7 +342,7 @@ class ResPartner(models.Model):
     fiscal_country_codes = fields.Char(compute='_compute_fiscal_country_codes')
     fiscal_country_group_codes = fields.Json(compute='_compute_fiscal_country_group_codes')
     partner_vat_placeholder = fields.Char(compute='_compute_partner_vat_placeholder')
-    duplicate_bank_partner_ids = fields.Many2many(related="bank_ids.duplicate_bank_partner_ids")
+    duplicate_bank_partner_ids = fields.Many2many('res.partner', compute='_compute_duplicate_bank_partner_ids')
 
     global_location_number = fields.Char(string="GLN", help="Global Location Number")
 
@@ -1039,6 +1039,11 @@ class ResPartner(models.Model):
 
             partner.partner_vat_placeholder = placeholder
 
+    @api.depends('bank_ids.duplicate_bank_partner_ids')
+    def _compute_duplicate_bank_partner_ids(self):
+        for partner in self:
+            partner.duplicate_bank_partner_ids = partner.bank_ids.duplicate_bank_partner_ids
+
     @api.depends('country_id')
     def _compute_company_registry_placeholder(self):
         """ Provides a dynamic placeholder on the company registry field for countries that may need it.

--- a/addons/account/models/res_partner_bank.py
+++ b/addons/account/models/res_partner_bank.py
@@ -8,6 +8,7 @@ from odoo import _, api, fields, models
 from odoo.exceptions import UserError, ValidationError
 from odoo.tools import SQL
 from odoo.tools.image import image_data_uri
+from odoo.addons.account.tools import format_account_number, validate_iban, validate_clabe
 
 
 class ResPartnerBank(models.Model):
@@ -18,27 +19,21 @@ class ResPartnerBank(models.Model):
         'account.journal', 'bank_account_id', domain=[('type', '=', 'bank')], string='Account Journal', readonly=True,
         check_company=True,
         help="The accounting journal corresponding to this bank account.")
-    has_iban_warning = fields.Boolean(
-        compute='_compute_display_account_warning',
-        help='Technical field used to display a warning if the IBAN country is different than the holder country.',
-        store=True,
+    has_phishing_warnings = fields.Boolean(compute='_compute_has_phishing_warnings', store=True)
+    phishing_warnings = fields.Json(
+        compute='_compute_phishing_warnings',
+        help="Technical field used to display warnings if there's a potential phishing risk",
     )
-    partner_country_name = fields.Char(related='partner_id.country_id.name')
-    has_money_transfer_warning = fields.Boolean(
-        compute='_compute_display_account_warning',
-        help='Technical field used to display a warning if the account is a transfer service account.',
-        store=True,
-    )
-    money_transfer_service = fields.Char(compute='_compute_money_transfer_service_name')
     partner_supplier_rank = fields.Integer(related='partner_id.supplier_rank')
     partner_customer_rank = fields.Integer(related='partner_id.customer_rank')
     related_moves = fields.One2many('account.move', inverse_name='partner_bank_id')
+    account_type = fields.Selection(selection_add=[('iban', 'IBAN'), ('clabe', 'CLABE')])
 
     # Add tracking to the base fields
-    bank_id = fields.Many2one(tracking=True)
+    bank_bic = fields.Char(tracking=True)
     active = fields.Boolean(tracking=True)
-    acc_number = fields.Char(tracking=True)
-    acc_holder_name = fields.Char(tracking=True)
+    account_number = fields.Char(tracking=True)
+    holder_name = fields.Char(tracking=True)
     clearing_number = fields.Char(tracking=True)
     partner_id = fields.Many2one(tracking=True)
     user_has_group_validate_bank_account = fields.Boolean(compute='_compute_user_has_group_validate_bank_account')
@@ -48,10 +43,22 @@ class ResPartnerBank(models.Model):
              'To protect yourself, always verify new bank account numbers, preferably by calling the vendor, as phishing '
              'usually happens when their emails are compromised. Once verified, you can activate the ability to send money.'
     )
-    currency_id = fields.Many2one(tracking=True)
     lock_trust_fields = fields.Boolean(compute='_compute_lock_trust_fields')
     duplicate_bank_partner_ids = fields.Many2many('res.partner', compute="_compute_duplicate_bank_partner_ids")
 
+    @api.model
+    def retrieve_account_type(self, account_number):
+        for validator, account_type in (
+            (validate_iban, 'iban'),
+            (validate_clabe, 'clabe'),
+        ):
+            try:
+                validator(self.env, account_number)
+                return account_type
+            except ValidationError:
+                pass
+        return super().retrieve_account_type(account_number)
+
     @api.constrains('journal_id')
     def _check_journal_id(self):
         for bank in self:
@@ -65,14 +72,14 @@ class ResPartnerBank(models.Model):
                 if not self.env.user.has_group('account.group_validate_bank_account'):
                     raise ValidationError(_('You do not have the right to trust or un-trust a bank account.'))
 
-    @api.depends('acc_number')
+    @api.depends('account_number')
     def _compute_duplicate_bank_partner_ids(self):
         id2duplicates = dict(self.env.execute_query(SQL(
             """
                 SELECT this.id,
                        ARRAY_AGG(other.partner_id)
                   FROM res_partner_bank this
-             LEFT JOIN res_partner_bank other ON this.acc_number = other.acc_number
+             LEFT JOIN res_partner_bank other ON this.account_number = other.account_number
                                              AND this.id != other.id
                                              AND other.active = TRUE
                  WHERE this.id = ANY(%(ids)s)
@@ -92,27 +99,63 @@ class ResPartnerBank(models.Model):
             duplicate_record = [x for x in duplicate_record if x]
             bank.duplicate_bank_partner_ids = self.env['res.partner'].browse(duplicate_record) if duplicate_record else False
 
-    @api.depends('partner_id.country_id', 'sanitized_acc_number', 'allow_out_payment', 'acc_type')
-    def _compute_display_account_warning(self):
-        for bank in self:
-            if bank.allow_out_payment or not bank.sanitized_acc_number or bank.acc_type != 'iban':
-                bank.has_iban_warning = False
-                bank.has_money_transfer_warning = False
-                continue
-            bank_country = bank.sanitized_acc_number[:2]
-            bank.has_iban_warning = bank.partner_id.country_id and bank_country != bank.partner_id.country_id.code
-
-            bank_institution_code = bank.sanitized_acc_number[4:7]
-            bank.has_money_transfer_warning = bank_institution_code in bank._get_money_transfer_services()
-
-    @api.depends('sanitized_acc_number', 'allow_out_payment')
-    def _compute_money_transfer_service_name(self):
-        for bank in self:
-            if bank.sanitized_acc_number:
-                bank_institution_code = bank.sanitized_acc_number[4:7]
-                bank.money_transfer_service = bank._get_money_transfer_services().get(bank_institution_code, False)
-            else:
-                bank.money_transfer_service = False
+    @api.depends('allow_out_payment', 'account_type', 'sanitized_account_number', 'partner_id.country_id', 'country_id')
+    def _compute_phishing_warnings(self):
+        warnings_common_data = {
+            'level': 'danger',
+            'action_text': self.env._("Learn more"),
+            'action': {
+                'type': 'ir.actions.act_url',
+                'url': 'https://www.odoo.com/documentation/latest/applications/finance/accounting/payables/pay/trusted_accounts.html',
+                'target': 'new',
+            },
+        }
+        code_to_country_id = dict(self.env['res.country']._read_group(
+            domain=[('code', 'in', iban_accounts.mapped(lambda account: account.sanitized_account_number[:2]))],
+            groupby=['code'],
+            aggregates=['id:recordset'],
+        )) if (iban_accounts := self.filtered(lambda account: account.account_type == 'iban')) else {}
+        for account in self:
+            warnings = {}
+            if not account.allow_out_payment and account.sanitized_account_number:
+                if account.account_type == 'iban' and (bank_institution_code := account.sanitized_account_number[4:7]) and (money_transfer_service := account._get_money_transfer_services().get(bank_institution_code)):
+                    warnings['service_not_bank'] = {
+                        'message': self.env._(
+                            "Phishing risk: %(money_transfer_service)s is a money transfer service and not a bank. Double check if the account can be trusted by calling the vendor.",
+                            money_transfer_service=money_transfer_service,
+                        ),
+                        **warnings_common_data,
+                    }
+
+                labels_and_countries = [
+                    (self.env._("IBAN"), account.account_type == 'iban' and code_to_country_id.get(account.sanitized_account_number[:2])),
+                    (self.env._("CLABE"), account.account_type == 'clabe' and self.env.ref('base.mx')),
+                    (self.env._("Account Holder"), account.partner_id.country_id),
+                    (self.env._("Bank Account"), account.country_id),
+                ]
+                for idx, (label1, country1) in enumerate(labels_and_countries, 1):
+                    if 'countries_mismatch' in warnings:
+                        break
+                    for label2, country2 in labels_and_countries[idx:]:
+                        if country1 and country2 and country1 != country2:
+                            warnings['countries_mismatch'] = {
+                                'message': self.env._(
+                                    "Phishing risk: %(label1)s is in %(country1)s but %(label2)s is in %(country2)s",
+                                    label1=label1,
+                                    country1=country1.name,
+                                    label2=label2,
+                                    country2=country2.name,
+                                ),
+                                **warnings_common_data,
+                            }
+                            break
+
+            account.phishing_warnings = warnings
+
+    @api.depends('phishing_warnings')
+    def _compute_has_phishing_warnings(self):
+        for account in self:
+            account.has_phishing_warnings = bool(account.phishing_warnings)
 
     def _get_money_transfer_services(self):
         return {
@@ -121,7 +164,7 @@ class ResPartnerBank(models.Model):
             '974': 'PPS EU SA',
         }
 
-    @api.depends('acc_number')
+    @api.depends('account_number')
     @api.depends_context('uid')
     def _compute_user_has_group_validate_bank_account(self):
         user_has_group_validate_bank_account = self.env.user.has_group('account.group_validate_bank_account')
@@ -282,20 +325,28 @@ class ResPartnerBank(models.Model):
                 vals['allow_out_payment'] = False
 
         for vals in vals_list:
-            if (partner_id := vals.get('partner_id')) and (acc_number := vals.get('acc_number')):
-                archived_res_partner_bank = self.env['res.partner.bank'].search([('active', '=', False), ('partner_id', '=', partner_id), ('acc_number', '=', acc_number)])
+            if vals.get('account_number'):
+                vals['account_number'] = format_account_number(self.env, vals['account_number'])
+
+            if (partner_id := vals.get('partner_id')) and (account_number := vals.get('account_number')):
+                archived_res_partner_bank = self.env['res.partner.bank'].search([('active', '=', False), ('partner_id', '=', partner_id), ('account_number', '=', account_number)])
                 if archived_res_partner_bank:
-                    raise UserError(_("A bank account with Account Number %(number)s already exists for Partner %(partner)s, but is archived. Please unarchive it instead.", number=acc_number, partner=archived_res_partner_bank.partner_id.name))
+                    raise UserError(_("A bank account with Account Number %(number)s already exists for Partner %(partner)s, but is archived. Please unarchive it instead.", number=account_number, partner=archived_res_partner_bank.partner_id.name))
 
         res = super().create(vals_list)
         res._check_allow_out_payment()
         for account in res:
+            partner = account.partner_id
+            msg = _("Partner set to %s", partner._get_html_link(title=f"#{partner.display_name}"))
+            account._message_log(body=msg)
             msg = _("Bank Account %s created", account._get_html_link(title=f"#{account.id}"))
-            account.partner_id._message_log(body=msg)
+            partner._message_log(body=msg)
         return res
 
     def write(self, vals):
         # EXTENDS base res.partner.bank
+        if vals.get('account_number'):
+            vals['account_number'] = format_account_number(self.env, vals['account_number'])
         # Track and log changes to partner_id, heavily inspired from account_move
         account_initial_values = defaultdict(dict)
         # Get all tracked fields (without related fields because these fields must be managed on their own model)
@@ -323,7 +374,7 @@ class ResPartnerBank(models.Model):
             # If we were on a trusted account, we only allow changes if the account is moving to untrusted.
             should_allow_changes = self.env.su or ('allow_out_payment' in vals and vals['allow_out_payment'] is False)
 
-        lock_fields = {'acc_number', 'sanitized_acc_number', 'partner_id', 'acc_type'}
+        lock_fields = {'account_number', 'sanitized_account_number', 'partner_id', 'account_type'}
         if not should_allow_changes and any(
             account[fname] != account._fields[fname].convert_to_record(
                 account._fields[fname].convert_to_cache(vals[fname], account),
@@ -353,20 +404,13 @@ class ResPartnerBank(models.Model):
                     initial_values['partner_id']._message_log(body=msg, tracking_value_ids=tracking_value_ids)
         return res
 
-    def unlink(self):
-        # EXTENDS base res.partner.bank
-        for account in self:
-            msg = _("Bank Account %(link)s with number %(number)s archived", link=account._get_html_link(title=f"#{account.id}"), number=account.acc_number)
-            account.partner_id._message_log(body=msg)
-        return super().unlink()
-
     @api.model
     def default_get(self, fields):
-        if 'acc_number' not in fields:
+        if 'account_number' not in fields:
             return super().default_get(fields)
 
         # When create & edit, `name` could be used to pass (in the context) the
         # value input by the user. However, we want to set the default value of
-        # `acc_number` variable instead.
-        default_acc_number = self.env.context.get('default_acc_number', False) or self.env.context.get('default_name', False)
-        return super(ResPartnerBank, self.with_context(default_acc_number=default_acc_number)).default_get(fields)
+        # `account_number` variable instead.
+        default_account_number = self.env.context.get('default_account_number', False) or self.env.context.get('default_name', False)
+        return super(ResPartnerBank, self.with_context(default_account_number=default_account_number)).default_get(fields)
